### DISCUSSION TITLE
Possible to capture a flat background colour with toBlob

### DISCUSSION THREAD

**Discussion by ghost** [google/model-viewer#3345] [Q&A]
Hello All,
I am just beginning to explore model-viewer and was wondering if someone could help me with a question I have related to capturing the model viewer canvas.

I am using the built-in ["toBlob"](https://modelviewer.dev/docs/index.html#entrydocs-loading-methods-toBlob) method of model-viewer to achieve this but no matter what I try, I cannot get the flat background-colour that I have set for the model-viewer to get captured.

--------------------------------------------------------------------
THE PROBLEM:

This is what I am doing:
1. Users can update the background colour of the viewer using an HTML colour input element. So I am setting the "background-color" on the model-viewer using this input value. This indeed updates the background-colour and looks fine in the viewer.
![image](https://user-images.githubusercontent.com/70152396/161943498-79bf4429-ae41-4cfd-8b8b-2a2b227c2d6c.png)

2. The user then clicks a button to capture a screenshot of what they see on their screen but the output is missing the background-color they had set earlier. The image attached shows a white background but it is transparent.
![image](https://user-images.githubusercontent.com/70152396/161943613-9cc179cd-3760-48bc-b67c-7787e133c806.png)

Now this is not the case if I add a skybox-image but I really don't want to do that and just use the neutral lighting. Is there a way to achieve this?

--------------------------------------------------------------------

WHAT I HAVE TRIED SO FAR:

Before I explain what I have tried so far, I believe this is due to the fact that setting "background-color" CSS property only updates what it says and never actually fills the canvas with that colour. There might be a straight forward way to do this that I am not familiar with.

1. I tried making another canvas and filling it with the selected colour by the user, then drawing this captured image on top of it to generate the same result. But this would definitely need more computation and memory, specially on Mobile device. Not a bad solution overall in my opinion but looking for a better options.
2. Using a flat skybox-image and then tinting the image itself at runtime to generate the bakground. The obvious problem with this is that it starts to affect the lighting in the scene, making it a non usable solution for me.

**Comment by elalish** [comment]
Indeed, the canvas is transparent and layered on top of the CSS background, so all you get is the transparent canvas pixels from `toBlob`. Generally this is fine, as you can simply put the poster image on top of the same CSS background and get the same effect. Otherwise you may need an image editing program (possibly web/canvas based) to do the merge. You could also look into the screenshot function in Puppeteer.

**Comment by ghost** [comment]
Thank you for confirming that.

So, is there no way to set a "Clear colour" on the camera itself without
affecting the scene lighting? Except, the obvious skybox-image?
Essentially, something that allows me to fill in those transparent pixels
with anything that I provide but not affecting the lighting. A callback per
frame that we could execute before the model gets drawn on the canvas could
prove useful too.

**Reply by elalish** [reply]
It's certainly possible if you hack down into our three.js renderer. This low-level control isn't really what our API aims to expose though. What's your use case?

**Comment by ghost** [comment]
That's what I was aiming to do next (threejs tweaks) if no existing
solution was available out of the box for now.

I am working on a review tool, that captures the model viewer frame and
allows the reviewer to draw on top of it + leave comments. I am currently
using a canvas to merge the model viewer frame with the background, then
displaying it for editing purposes. Although it will start to get a bit
complicated once we have gradient backgrounds and images, specially because
the background colour can be edited by the reviewer for better visibility.

**Comment by tanaydimri** [comment]
I had to delete that github account for some security reasons.
Commenting from my permanent account now.

**Comment by seadeep42** [comment]
Hi,
I hope this might help. I had a similar use case and was able to use the below code to set the canvas clear color after model-viewer load.

```
    const canvas = document.getElementById("model-viewer").shadowRoot.querySelector('.slot.canvas #webgl-canvas');
    if (canvas.getContext('webgl2')) {
      const gl = canvas.getContext('webgl2');
      gl.clearColor(r, g, b, 1);
    } else if (canvas.getContext('webgl')) {
      const gl = canvas.getContext('webgl');
      gl.clearColor(r, g, b);
    }
```

**Reply by tanaydimri** [reply]
Hey @chaitanya-deep,
Thank you for the suggestion.
Unfortunately, this is something I have been planning to avoid for two main reasons:

1. I usually refrain myself from accessing private elements as a general rule. Not sure if this one is applicable here because the shadow root is open to be edited. Have you had any issues yet with the Major version updates to the model-viewer changing these private elements? If not and if it's a common practise to modify canvas within the model-viewer component. This solution might just work for me.
2. I am unsure if this would interfere with any of the existing logics of model-viewer. Especially the skybox-image rendering. Although Clear Colour would only draw on the "empty pixels" but this might not work when I am drawing an entire image and not just setting the Clear Colour.

Would like to hear your thoughts.

**Reply by elalish** [reply]
It's definitely outside of our public API, so your concerns are valid. That said, this is probably not super dangerous. But definitely you'll want to do some testing before upgrading to each version.

**Reply by seadeep42** [reply]
1. I started using model viewer recently and there haven't been major versions since then. But I was prepared to verify before updates since the only thing that might break is the query selector incase the id/path is changed which I was prepared for.
2. So far for my usage, it didn't seem to interfere. I use it to change the background color of video download from the browser. https://celestar.euclidxr.com/video/968c2f40-770c-411e-8dc1-633b0bcbd203

**Reply by tanaydimri** [reply]
Thanks to both of you for insights on this.
I decided to move forward with the approach of merging the viewer with no background for now and merging it with a background of my choice. I majorly chose this approach because I saw the requirement for a system like for other cases too within the tool.

If anyone is ever looking for a solution on this thread, this is what I ended up doing for my use case:

```
export class ImageMerger 
{
    constructor(width=1000, height=1000) 
    {
        this._mergerCanvas = document.createElement("canvas");
        this._mergerCanvas.id = "mergeCanvas";
        this._mergerCanvas.width = width;
        this._mergerCanvas.height = height;
        this._mergerCanvasCtx = this._mergerCanvas.getContext("2d");

        // Uncomment for debugging purposes
        // document.body.appendChild(this._mergerCanvas);
    }

    dataURLtoBlob(dataurl) {
        var arr = dataurl.split(','), mime = arr[0].match(/:(.*?);/)[1],
            bstr = atob(arr[1]), n = bstr.length, u8arr = new Uint8Array(n);
        while(n--){
            u8arr[n] = bstr.charCodeAt(n);
        }
        return new Blob([u8arr], {type:mime});
    }

    async mergeImages(canvasClearColour = "rgba(0, 0, 0, 0)", modelViewportCapture) 
    {
        // Fill the background clear colour with the provided Canvas Clear Colour first
        this._mergerCanvasCtx.fillStyle = canvasClearColour;
        this._mergerCanvasCtx.fillRect(0, 0, 1000, 1000);

        // TODO: Any images that need to be layered could be drawn here.
        // Maybe accept an array from "modelViewportCapture" parameter?

        // Drawing the provided image on top of the canvas
        const tempImage = new Image();
        tempImage.src = modelViewportCapture;
        await tempImage.decode();

        // Calculating the new width and height to ensure that drawn image is always
        // in the center of the canvas
        const newImageDrawWidth = (this._mergerCanvas.width/2 - tempImage.width/2);
        const newImageDrawHeight = (this._mergerCanvas.height/2 - tempImage.height/2);
        this._mergerCanvasCtx.drawImage(tempImage, newImageDrawWidth, newImageDrawHeight);

        const canvasData = await this._mergerCanvas.toDataURL();
        const imageBlob = this.dataURLtoBlob(canvasData);
        const canvasImageUrl = URL.createObjectURL(imageBlob);

        return canvasImageUrl;
    }
}
```
